TRB_label_assist/models.py

```diff
@@ -378,13 +405,19 @@ def _distance_helper(self, tripa, tripb, loc_type):
         
             copied from the Similarity class on the e-mission-server. 
```

Comment:
Now that we are no longer on a custom branch, can't the distance calculation in `e-mission-server` be re-used here? Why do we need copy-pasted code? I am generally again copy-pasting to support DRY. Since this change has already had multiple revisions, I am OK with deferring this to the next PR, but I want to make sure that it is not forgotten.

Reply:
We are using the `e-mission-server` :

```
  dist= ecc.calDistance([pta_lon,pta_lat],[ptb_lon,ptb_lat])     
```                              

ecc here is on `e-mission-server` :
```
  import emission.core.common as ecc
```

I'll remove that old comment from Hannah.

Reply:
While predicting in the `greedy_similarity_binning.py` on `e-mission-server`,  the flow goes like :

 `predict` -> `_nearest_bin` ->`similar` ( in `e-mission-server/emission/analysis/modelling/similarity/similarity_metric.py`) ->`similarity` ( in `e-mission-server/emission/analysis/modelling/similarity/od_similarity.py`) -> `ecc.calDistance`

Reply:
I think that you have misunderstood the comment. The comment is not about `ecc.calDistance` - it is clear that `calDistance` is from 
This says that `_distanceHelper` is copied from e-mission-server, which is it

```
$ grep -r distance_helper emission/
emission//analysis/modelling/tour_model_first_only_orig/similarity.py:            if not self.distance_helper(a, b):
emission//analysis/modelling/tour_model_first_only_orig/similarity.py:    def distance_helper(self, a, b):
emission//analysis/modelling/tour_model/similarity.py:            if not self.distance_helper(a, b):
emission//analysis/modelling/tour_model/similarity.py:    def distance_helper(self, a, b):
emission//incomplete_tests/TestSimilarity.py:                    self.assertTrue(sim.distance_helper(b,c))
```

However, the implementation does seem to be a bit different, and I don't see a function with this name in trip_model. There must be an equivalent in trip_model.

Reply:
Other than `ecc.calDistance`, currently  pre-processing is done  in the form of coordinate extraction in `_distanceHelper`:

```
        #tripa is taken from the test datframe. 
        #tripb is taken from the stored bin list.
        
        pta_lat = tripa[[loc_type + '_lat']]
        pta_lon = tripa[[loc_type + '_lon']]
        if loc_type == 'start':
            ptb_lat = tripb[1]
            ptb_lon = tripb[0]
        elif loc_type == 'end':
            ptb_lat = tripb[3]
            ptb_lon = tripb[2]
```


We do have `extract_features` function ( at `e-mission-server/emission/analysis/modelling/similarity/od_similarity.py`) that extracts latitude and longitude of trips, but it works just for `Entry` type data ( since data frames are not used in `e-mission-server`). 
Here, in this repo, we additionally have `tripa` ( the test trip),which is row of data frame received from the `predict` function. 


>There _must_ be an equivalent in trip_model to calculate distances between trips, which we should reuse here.

For the reason above, there isn't. However, this is  what we can do  to use the `_nearest_bin` function ( `e-mission-server/emission/analysis/modelling/trip_model/greedy_similarity_binning.py`)  which is closest to `_distance_helper` :

1. convert `tripa` ( the test trip) to entry type data  using  `df_row_to_entry` ( in `e-mission-server/emission/storage/timeseries/builtin_timeseries.py`). 
2. Pass this entry to the `_nearest_bin` function.

Let me know if this works, and I'll test this.

Reply:
From https://github.com/e-mission/e-mission-eval-private-data/pull/37#discussion_r1398304851

> Since this change has already had multiple revisions, I am OK with deferring this to the next PR, but I want to make sure that it is not forgotten.

I am fine with returning to this later. However, eventually, we should simplify the codebase to either use only dataframes, or use only entries, or, if we are going to support some level of mix and match, have the utility functions support both combinations.